File: 0098-validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def listTree(self, root, l):
        if root.left:
            self.listTree(root.left, l)
        l.append(root.val)
        if root.right:
            self.listTree(root.right, l)
        return    
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root == None:
            return True
        
        l = []
        self.listTree(root,l)
        
        for i in xrange(len(l)-1):
            if l[i] >= l[i+1]:
                return False
        
        return True
        

File: 0098-validate-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool helper(TreeNode* root, long long L, long long R) {
        if (root == nullptr) return true;

        if (L < root->val && root->val < R) {
            return helper(root->left, L, root->val) && helper(root->right, root->val, R);
        }

        return false;
    }
    bool isValidBST(TreeNode* root) {
        long long L = LLONG_MIN, R = LLONG_MAX;
        return helper(root, L, R);
    }
};

